### PR TITLE
[http-client-csharp] fix: use correct custom ctor in model factory

### PR DESCRIPTION
**Pull request by jorgerangel-msft**
This PR fixes an issue where the model factory method for a model was using the incorrect full constructor when the full constructor was suppressed and customized.

fixes: https://github.com/microsoft/typespec/issues/4830

### CODE REVIEW

File: packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/Providers/ModelFactoryProvider.cs [hunk 1]
```diff
@@ -79,14 +79,39 @@ protected override MethodProvider[] BuildMethods()
...
                     continue;
 
-                var modelCtor = modelProvider.FullConstructor;
+                var fullConstructor = modelProvider.FullConstructor;
+                var binaryDataParam = fullConstructor.Signature.Parameters.FirstOrDefault(p => p.Name.Equals(AdditionalBinaryDataParameterName));
+
+                // Use a custom constructor if the generated full constructor was suppressed or customized
+                if (!modelProvider.Constructors.Contains(fullConstructor))
+                {
+                    foreach (var constructor in modelProvider.CanonicalView?.Constructors ?? [])
```
Comment (JoshLove-msft) [inline packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/Providers/ModelFactoryProvider.cs:88]: Why do we need `?` for the CanonicalView. I thought this was not nullable.

Comment (jorgerangel-msft) [reply packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/Providers/ModelFactoryProvider.cs]: fixed.

File: packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/Providers/ModelFactoryProvider.cs [hunk 2]
```diff
@@ -79,14 +79,39 @@ protected override MethodProvider[] BuildMethods()
...
                     continue;
 
-                var modelCtor = modelProvider.FullConstructor;
+                var fullConstructor = modelProvider.FullConstructor;
+                var binaryDataParam = fullConstructor.Signature.Parameters.FirstOrDefault(p => p.Name.Equals(AdditionalBinaryDataParameterName));
+
+                // Use a custom constructor if the generated full constructor was suppressed or customized
+                if (!modelProvider.Constructors.Contains(fullConstructor))
+                {
+                    foreach (var constructor in modelProvider.CanonicalView.Constructors ?? [])
```
Comment (JoshLove-msft) [inline packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/Providers/ModelFactoryProvider.cs:88]: Do we still need the null coalesce?

Comment (jorgerangel-msft) [reply packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/Providers/ModelFactoryProvider.cs]: no we don't. sorry it's that time of the day now 😄

**Pull request merged**